packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js

```diff
@@ -0,0 +1,301 @@
+import locationValidator from '../../helpers/locationValidator'
+import withKeyZoom from '../../../withKeyZoom'
+
+function preventDefault (e) {
+  e.preventDefault()
+}
+
+const DEFAULT_CELL_WIDTH = 1200
+const DEFAULT_CELL_HEIGHT = 1000
+const DEFAULT_CELL_STYLE = {
+  stroke: '#fff',
+  strokeWidth: '4',
+  fill: '#000'
+}
+const DEFAULT_GRID_COLUMNS = 3
+const DEFAULT_GRID_ROWS = 3
+
+function storeMapper (stores) {
+  const {
+    setOnZoom,
+    setOnPan
+  } = stores.classifierStore.subjectViewer
+  
+  const {
+    active: activeWorkflow
+  } = stores.classifierStore.workflows
+  
+  const workflowConfig = activeWorkflow && activeWorkflow.configuration || {}
+  
+  const cellWidth = workflowConfig.cell_width || DEFAULT_CELL_WIDTH
+  const cellHeight = workflowConfig.cell_height || DEFAULT_CELL_HEIGHT
+  const cellStyle = workflowConfig.cell_style || DEFAULT_CELL_STYLE
+  const gridColumns = workflowConfig.grid_columns || DEFAULT_GRID_COLUMNS
+  const gridRows = workflowConfig.grid_rows || DEFAULT_GRID_ROWS
+
+  return {
+    cellWidth,
+    cellHeight,
+    cellStyle,
+    gridColumns,
+    gridRows,
+    setOnZoom,
+    setOnPan
+  }
+}
+
+class SubjectGroupViewerContainer extends React.Component {
+  constructor () {
+    super()
+    this.dragMove = this.dragMove.bind(this)
+    this.groupViewer = React.createRef()
+    this.scrollContainer = React.createRef()
+
+    this.state = {
+      images: [],
+      panX: 0,
+      panY: 0,
+      zoom: 1,
+    }
+  }
+
+  componentDidMount () {
+    this.onLoad()
+
+    // Listen for pan and zoom actions outside of this component.
+    // i.e. zoom in/out actions from the iamge controls. 
+    this.props.setOnPan(this.onPanViaExternalControls.bind(this))
+    this.props.setOnZoom(this.onZoomViaExternalControls.bind(this))
+    
+    this.scrollContainer.current
+    && this.scrollContainer.current.addEventListener('wheel', preventDefault)
+  }
+  
+  componentWillUmount () {
+
+    this.setOnPan(() => true)
+    this.setOnZoom(() => true)
+    
+    this.scrollContainer.current
+    && this.scrollContainer.current.removeEventListener('wheel', preventDefault)
+  }
+
+  fetchImage (url) {
+    const { ImageObject } = this.props
+    return new Promise((resolve, reject) => {
+      const img = new ImageObject()
+      img.onload = () => resolve(img)
+      img.onerror = reject
+      img.src = url
+      return img
+    })
+  }
+
+  dragMove (event, difference) {
+    this.onDrag && this.onDrag(event, difference)
+  }
+
+  async preload () {
+    const { subject } = this.props
+    if (subject && subject.locations) {
+      // TODO: Validate for allowed image media mime types
+      
+      const imageUrls = subject.locations.map(obj => Object.values(obj)[0])
+      const images = await Promise.all(
+        imageUrls.map(url => this.fetchImage(url))
+      )
+      
+      this.setState({ images })
+      return images
+    }
+    return {}
+  }
+
+  async getImageSize () {
```

Comment:
Also, I don't see anything inside this function which would make it asynchronous.

Reply:
This should be completely renamed and updated, since the _"get the sizes of the visual elements so we can calculate how we can display them"_ function has mutated greatly from the comparable SingleImageViewer code. I originally tried to keep the SubjectGroupViewer and SingleImageViewer as similar as possible, but I think the dev for the SGV has sailed waaay past that point already.

To be precise: this should be named `getGridSize()` and doesn't need to be async.